Transient poisson multimaterial example

flothesof opened the pull request:
Hi, as discussed in #900 and #898, I would like to submit this example.

I’ve developed it under Windows and had a number of issues during development of this example.
Most importantly, the documentation generation is broken under windows (and installation from source also).
So I couldn’t really see how the example is rendered.

What do you think?

Cheers,
Florian

rc commented:
Thanks, very nice!

Just a few notes:
- Format also the docstring to 80 chars line width.
- LaTeX chokes on the second equation - replace align -> aligned
- Add signs to terms in the equation :)
- The integral order two is OK as long as the material parameters are constant per element - it seems to be so here, right?
- Maybe replace `ls.scipy_direct` by `ls.auto_direct`, which can use MUMPS, if available. Also set a few options for linear problems:
```
    'ls' : ('ls.auto_direct', {
        # Reuse the factorized linear system from the first time step.
        'use_presolve' : True,
        # Speed up the above by omitting the matrix digest check used normally
        # for verification that the current matrix corresponds to the
        # factorized matrix stored in the solver instance. Use with care!
        'use_mtx_digest' : False,
    }),
```
is more than 4x faster on my laptop...

- Set `'i_max': 1,` in `nls.newton`.

In your application, do you plan to store a plot a time-history of the solution at some point in the domain? If yes, it would be nice also here.

flothesof commented:
Thank you for your comments.
I’ll try to come up with an improved version of the script.

Regarding the time-history, I am indeed using a probe for extraction at every time step using step_hook
I can try to integrate this because indeed this would be even nicer.

rc commented:
Thanks! I can do the formatting for you (well, my editor can ;)) - can I push to your branch? IMO github allows that.

flothesof commented:
Hi, yes, you can push to my branch! The little checkbox allowing this is checked :)

rc commented:
I have pushed my changes, so do not forget to fetch them :)

flothesof commented:
I have removed the call to plt.show (I found it a bit too convoluted to add the `define()` function just for this) and added the example to the list of testable examples.

rc commented:
Thanks, LGTM! I have just added a custom view for the gallery.

rc commented on sfepy/tests/test_declarative_examples.py:
```diff
@@ -25,6 +25,7 @@ def inedir(filename):
     'diffusion/sinbc.py',
     'diffusion/time_advection_diffusion.py',
     'diffusion/time_poisson.py',
+    'diffusion/time_heat_equation_multi_material.py',
```
Put above time_poisson.py to preserve alphabetical order?

flothesof replied:
Alright!

rc commented:
OK, let's merge it!
